pkg/utils/k8sutil.go

```diff
@@ -302,7 +316,28 @@ func DeleteKafkaTriggerCustomResource(kubelessClient versioned.Interface, kafkaT
 	return nil
 }
 
-// GetKafkaTriggerCustomResource will get CronJobTrigger custom resource object
+// PatchKafkaTriggerCustomResource patches changes to the KafkaTrigger CRD object
+func PatchKafkaTriggerCustomResource(kubelessClient versioned.Interface, original, patched *kubelessApi.KafkaTrigger) error {
+	curJSON, err := json.Marshal(original)
+	if err != nil {
+		return err
+	}
+	modJSON, err := json.Marshal(patched)
+	if err != nil {
+		return err
+	}
+	patch, err := jsonmergepatch.CreateThreeWayJSONMergePatch(curJSON, modJSON, curJSON)
+	if err != nil {
+		return err
+	}
+	if len(patch) == 0 || string(patch) == "{}" {
+		return nil
+	}
+	_, err = kubelessClient.KubelessV1beta1().KafkaTriggers(original.Namespace).Patch(original.Name, types.MergePatchType, patch)
```

Comment:
Just repeating earlier comment:  These patches need to be applied against a particular resource version (with retry on conflict detection) - otherwise we can race against another change between reading the "original" version and applying the patch.

Reply:
So both the update/patch has to be done against the latest resource version? Do you recall any core controllers of kube-controller-manger does this retry on conflict detection logic. All I see is they just do update with out retry mostly.

Reply:
See all the calls to `k8s.io/client-go/util/retry.RetryOnConflict`.  I presume the other Update calls just propagate the conflict error up the stack, at worst doing a full requeue and error retry later.

The example of a Patch I've just looked at (`pkg/controller/volume/expand/cache.UpdatePVSize`) does a patch that is idempotent, and generated+used within a single simple function, so the call to `CreateTwoWayMergePatch()` is just an expensive way to calculate a tiny/known patch object.  This use of `Patch()` is fine since the resulting patch doesn't actually depend on the original object anymore.  If you have another specific example in mind, then I'd be happy to look into it.

I needed to stare at the kubeless code a lot longer to work out if the kubeless patches where also idempotent.  In particular, things like having time-separated calls to `previousFunction, err := utils.GetFunctionCustomResource(kubelessClient, ...)` and `updatedFunction, err := getFunctionDescription(cli, ...)` makes me very nervous .. but it seems `cli` isn't actually used(?) in the latter.  Staring harder at the `getFunctionDescription()` object, some specific constructs like the loop that incrementally appends volumes/volumeMounts for each secret will clearly produce the wrong patch if something else also adds a volume between when `previousFunction` was fetched and `Patch()` is called.
My point is that concluding that this use of Patch was _not_ safe took about 15mins of looking for specific issues.  This is not a good pattern to follow for robust code (and also actually wrong in this case).

Reply:
TL;DR; I will revert the last commit that does 3-way merge patches and keep the update.

> things like having time-separated calls to previousFunction, err := utils.GetFunctionCustomResource(kubelessClient, ...) and updatedFunction, err := getFunctionDescription(cli, ...) makes me very nervous .. 

Agree. getFunctionDescription has grown big. And now that we have deployment sepc etc as part of Function spec, it's getting complex. But its CLI only change. If some object changes which cli execution in flight, cli fails. 

> My point with the comparison to UpdatePVSize is that concluding that kubeless' use of Patch was not safe took about 20mins of looking for specific issues, since the code between Get+Patch is quite lengthy and complex

Actually from controllers use of patch is pretty straightforward and quick. All controllers are doing is updating Finalizer.  Anyway, my intention was to avoid update on top of changed object resulting in `the object has been modified; please apply your changes to the latest version and try again`. I guess i will just keep the retry logic on update failure. All the core controllers simply perform updates. 

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L822

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L518

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/sync.go#L516

It also seems use of 3-way merge `Patch` may be overkill for controller. First, very likely all controller will be dealing is object meta and status. It really does not have same conflicts that updates to `Spec` would need. 

So to summarise, i will revert last commit. Update is what seems more logical for controller, but update alone is not sufficient but would need retry on resource version conflict.

Reply:
TL;DR; I will revert the last commit that does 3-way merge patches and keep the update.

> things like having time-separated calls to previousFunction, err := utils.GetFunctionCustomResource(kubelessClient, ...) and updatedFunction, err := getFunctionDescription(cli, ...) makes me very nervous .. 

Agree. getFunctionDescription has grown big. And now that we have deployment sepc etc as part of Function spec, it's getting complex. But its CLI only change. If some object changes which cli execution in flight, cli fails. 

> My point with the comparison to UpdatePVSize is that concluding that kubeless' use of Patch was not safe took about 20mins of looking for specific issues, since the code between Get+Patch is quite lengthy and complex

Actually from controllers use of patch is pretty straightforward and quick. All controllers are doing is updating Finalizer.  Anyway, my intention was to avoid update on top of changed object resulting in `the object has been modified; please apply your changes to the latest version and try again`. I guess i will just keep the retry logic on update failure. All the core controllers simply perform updates. 

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L822

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L518

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/sync.go#L516

It also seems use of 3-way merge `Patch` may be overkill for controller. First, very likely all controller will be dealing is object meta and status. It really does not have same conflicts that updates to `Spec` would need. 

So to summarise, i will revert last commit. Update is what seems more logical for controller, but update alone is not sufficient but would need retry on resource version conflict.